lib/cli/ui/prompt.rb

```diff
@@ -48,26 +49,29 @@ class << self
         # Free form question when the answer can be empty
         #   CLI::UI::Prompt.ask('What is your opinion on this question?', allow_empty: true)
         #
-        # Question with answers
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
+        # Multiple choice question with defined handlers
+        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python)) do |handler|
+        #     handler.add_option('rails')  { |selection| puts selection } => outputs "rails" if selected
+        #     handler.add_option('go')     { |selection| puts selection } => outputs "go" if selected
+        #     handler.add_option('ruby')   { |selection| puts selection } => outputs "ruby" if selected
+        #     handler.add_option('python') { |selection| puts selection } => outputs "python" if selected
         #
         #
-        def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true)
-          if (default && !allow_empty) || (options && (default || is_file))
+        def ask(question, default: nil, is_file: nil, allow_empty: true, &define_options)
+          if (default && !allow_empty) || (block_given? && (default || is_file))
             raise(ArgumentError, 'conflicting arguments')
           end
 
-          if default
-            puts_question("#{question} (empty = #{default})")
-          elsif options
-            puts_question("#{question} {{yellow:(choose with ↑ ↓ ⏎)}}")
-          else
-            puts_question(question)
-          end
-
           # Present the user with options
-          if options
-            resp = InteractiveOptions.call(options)
+          if block_given?
+            require 'cli/ui/prompt/options_handler'
```

Comment:
We normally autoload everything within these gems

Reply:
![image](https://user-images.githubusercontent.com/3074765/34583217-eb057582-f164-11e7-8b70-8db7acaeb1f8.png)

Reply:
It seemed like `autoload` is being deprecated? (although this was posted around 6 years ago) https://bugs.ruby-lang.org/issues/5653

Reply:
The reason it's being deprecated has to do with race conditions in multithreaded modes (not typically a major concern in CLI UIs). Even it it does go away, it's still a useful feature and we can reimplement it easily enough.

Reply:
More importantly, the project is currently autoloading everything. We can have a discussion about whether to switch to late-requires, but let's do that in a different PR: just follow the convention for now.